### PR TITLE
vsphere CSI Driver documentation

### PR DESCRIPTION
**Pull request by c-snell**

### CODE REVIEW

File: docs/partners/vmware/index.md [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+If there are no errors, the vSphere CSI Driver has been successfully deployed.
+
+### Validate the vSphere CSI Driver
+
+With the vSphere CSI Driver deployed, lets run through some tests to verify it is working correctly.
+
+!!! Important
+    The following steps will be using the example VM Storage Policy created at the beginning of this guide. If you do not have a Storage Policy available, refer to [Configuring a VM Storage Policy](#configuring_a_vm_storage_policy) before proceeding to the next steps.
+
+In this example, we will be deploying a stateful MongoDB application with 3 replicas. The persistent volumes deployed by the vSphere CSI Driver will be created using the VM Storage Policy and placed on a compatible vVOL datastore.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:391]: - VVOL lingo

File: docs/partners/vmware/index.md [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+
+##### Create a StorageClass
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:393]: I personally think the `StorageClass` should be part of the deployment process and not the "test & verify" phase of the guide.

File: docs/partners/vmware/index.md [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+
+Create the `StorageClass` that will be used for the MongoDB deployment. We will use the Storage Policy created earlier as an example. Make sure to change the Storage Policy name to match your setup.
+
+```markdown
+kubectl create -f-
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:398]: This needs to go.

File: docs/partners/vmware/index.md [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+    storageclass.kubernetes.io/is-default-class: "true"
+provisioner: csi.vsphere.vmware.com
+parameters:
+  storagepolicyname: "primera-default-profile"
+  fstype: ext4
+```
+
+Press **Enter** and **Ctrl-D**.
+
+Inspect the `StorageClass` to verify it was created successfully.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:417]: I find this a bit redundant, if `kubectl` doesn't return any errors, the resource will be created.

File: docs/partners/vmware/index.md [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+
+```markdown
+kubectl get sc sc-mongodb
+```
+
+##### Create and Deploy a MongoDB StatefulSet
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:423]: There's simply too much minutia here, can't we just use a Helm chart to achieve the same thing?

File: docs/partners/vmware/index.md [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+NAME                                  STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+mongodb-persistent-storage-mongod-0   Bound    pvc-fd3994fb-a5fb-460b-ab17-608a71cdc337   50Gi       RWO            sc-mongodb      1m
+mongodb-persistent-storage-mongod-1   Bound    pvc-a3755dbe-210d-4c7b-8ac1-bb0607a2c537   50Gi       RWO            sc-mongodb      2m
+mongodb-persistent-storage-mongod-2   Bound    pvc-22bab0f4-8240-48c1-91b1-3495d038533e   50Gi       RWO            sc-mongodb      2m
+```
+
+To interact with the Mongo replica set, you can connect to the first node.
+
+```markdown
+kubectl exec -it mongod-0 -c mongod-container bash
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:556]: I believe you can exec into sts/mongod and a working Pod will be picked.

File: docs/partners/vmware/index.md [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+```
+
+Once logged into the instance, type **mongo**.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:559]: I also think we should highlight the fact that a persistent volume is attached in the filesystem hierarchy by illustrating it with `df <the path>`.

**Review by datamattsson — CHANGES_REQUESTED** [review]
A few nits.

File: docs/partners/vmware/index.md [hunk 8]
```diff
@@ -0,0 +1,584 @@
+Cloud Native Storage for vSphere
+
+# Overview
+
+Cloud Native Storage (CNS) for vSphere exposes vSphere storage and features to Kubernetes users and was introduced in vSphere 6.7 U3. CNS is made up of two parts, a Container Storage Interface (CSI) driver for Kubernetes used to provision storage on vSphere and the CNS Control Plane within vCenter allowing visibility to persistent volumes through the new CNS UI within vCenter.
+
+CNS fully supports Storage Policy-Based Management (SPBM) to provision volumes. SPBM is a feature of VMware vSphere that allows an administrator to match VM workload requirements against storage array capabilities, with the help of VM Storage profiles. This storage profile can have multiple array capabilities and data services, depending on the underlying storage you use. HPE Storage (HPE Primera, Nimble Storage, Nimble Storage dHCI, and HPE 3PAR) has the largest user base of vVols in the market, due to its simplicity to deploy and ease of use.
```
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/partners/vmware/index.md:7]: fixed

File: docs/partners/vmware/index.md [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+
+[TOC]
+
+### Deployment
+
+When considering to use block storage within Kubernetes clusters running on VMware, customers need to evaluate which data protocol (FC or iSCSI) is primarily used within their virtualized enviroment. This will help best determine which CSI Driver can be deployed within your Kubernetes clusters.
```
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/partners/vmware/index.md:13]: fixed

File: docs/partners/vmware/index.md [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+    The HPE CSI Driver for Kubernetes does not support N_Port ID Virtualization (NPIV).
+
+|  Protocol   | HPE CSI Driver | vSphere CSI Driver | 
+| ------ | :--------: | :---------: | 
+| FC | - | X |
+| iSCSI | X | X |
+
+#### Prerequisites
+
+This guide will cover the configuration and deployment of the vSphere CSI Driver. Cloud Native Storage for vSphere uses the VASA provider and Storage Policy Based Management (SPBM) to create First Class disks on supported arrays.
```
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/partners/vmware/index.md:27]: yes

File: docs/partners/vmware/index.md [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+
+[Deploying a Kubernetes Cluster on vSphere with CSI and CPI](https://cloud-provider-vsphere.sigs.k8s.io/tutorials/kubernetes-on-vsphere-with-kubeadm.html)
+
+!!! Note
+    The following is an simplified single-site configuration to demonstrate how to deploy the vSphere CPI and CSI Driver. Make sure to adapt the configuration to match your environment and needs.
+
+
+##### Create a CPI ConfigMap
+
+The following steps are only done on the master node.
```
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/partners/vmware/index.md:85]: This is per their documentation.

File: docs/partners/vmware/index.md [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+To create the taint, run the following command for each node in your cluster. 
+
+```markdown
+kubectl taint node <node_name> node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
+```
+
+Verify the taint has been applied to each node.
+
+```markdown
+kubectl describe nodes | egrep "Taints:|Name:"
```
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/partners/vmware/index.md:181]: While the json is possible, I like the simplicity of the original.

File: docs/partners/vmware/index.md [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+
+![Click Finish](img/profile7.png)
+
+Now that we have configured a Storage Policy, we can proceed with the deployment of the vSphere CSI Driver.
+
+#### Install the vSphere Cloud Provider Interface (CPI)
+
+This is adapted from the following tutorial, please read over to understand all of the vSphere, firewall and guest OS requirements.
+
+[Deploying a Kubernetes Cluster on vSphere with CSI and CPI](https://cloud-provider-vsphere.sigs.k8s.io/tutorials/kubernetes-on-vsphere-with-kubeadm.html)
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:88]: Could we add a bullet here to make it a bit more emphasized, like this?

- [Deploying a Kubernetes Cluster on vSphere with CSI and CPI](https://cloud-provider-vsphere.sigs.k8s.io/tutorials/kubernetes-on-vsphere-with-kubeadm.html)

File: docs/partners/vmware/index.md [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+clusterrole.rbac.authorization.k8s.io/system:cloud-controller-manager created
+
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/master/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml
+clusterrolebinding.rbac.authorization.k8s.io/system:cloud-controller-manager created
+
+kubectl apply -f https://github.com/kubernetes/cloud-provider-vsphere/raw/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml
+serviceaccount/cloud-controller-manager created
+daemonset.extensions/vsphere-cloud-controller-manager created
+service/vsphere-cloud-controller-manager created
+```
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:218]: This is deprecated according to https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager

We should not endorse installing deprecated software, it now points to a single manifest (which makes sense).

File: docs/partners/vmware/index.md [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+
+!!! Note
+    If you happen to make an error with the **vsphere.conf**, simply delete the CPI components and the `ConfigMap`, make any necessary edits to the **vsphere.conf** file, and reapply the steps above.
+
+Now that the CPI is installed, we can proceed with deploying the vSphere CSI Driver.
+
+#### Install vSphere Container Storage Interface (CSI) Driver
+
+The following has been adapted from the vSphere CSI Driver installation guide. Refer to [https://vsphere-csi-driver.sigs.k8s.io/driver-deployment/installation.html](https://vsphere-csi-driver.sigs.k8s.io/driver-deployment/installation.html) for additional information on how to deploy the vSphere CSI Driver.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:241]: Can we make this consistent with #install_the_vsphere_cloud_provider_interface_cpi ? It doesn't matter if use this way of embedding the reference or the former.

File: docs/partners/vmware/index.md [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+
+Copy and paste the following:
+```markdown
+[Global]
+cluster-id = "csi-vsphere-cluster"
+
+[VirtualCenter "<IP or FQDN>"]
+insecure-flag = "true"
+user = "Administrator@vsphere.local"
+password = "VMware1!"
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:257]: I'm just curious where this password pattern stems from? If this is something we use commonly internally we should not expose it externally.

File: docs/partners/vmware/index.md [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+port = "443"
+datacenters = "<vCenter datacenter>"
+```
+
+##### Create a Kubernetes Secret for vSphere credentials
+
+Create a Kubernetes ``Secret` that will contain the configuration details to connect to your vSphere environment.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:264]: A back tick too many in `Secret`

File: docs/partners/vmware/index.md [hunk 18]
```diff
@@ -0,0 +1,451 @@
...
+
+```markdown
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.0/vsphere-7.0/vanilla/deploy/vsphere-csi-controller-deployment.yaml
+```
+
+vSphere CSI node `Daemonset`:
+
+```markdown
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.0/vsphere-7.0/vanilla/deploy/vsphere-csi-node-ds.yaml
+```
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:300]: Two things, there's a newer version 2.0.1 available. Also can we make this section a tabbed codeblock, one for 6.7U3 and one for 7.0? It seems the CSI driver supports both: https://github.com/kubernetes-sigs/vsphere-csi-driver/tree/master/manifests/v2.0.1

File: docs/partners/vmware/index.md [hunk 19]
```diff
@@ -0,0 +1,451 @@
...
+
+In this example, we will be deploying a stateful MongoDB application with 3 replicas. The persistent volumes deployed by the vSphere CSI Driver will be created using the VM Storage Policy and placed on a compatible vVol datastore.
+
+##### Create and Deploy a MongoDB Helm chart
+
+This is an example MongoDB chart using a StatefulSet.
+
+```markdown
+helm install mongodb \
+    --set architecture=replicaset,replicaSetName=mongod,replicaCount=3,auth.rootPassword=secretpassword,auth.username=my-user,auth.password=my-password,auth.database=my-database,persistence.storageClass=primera-default-sc,persistence.size=50Gi \
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:395]: If we omit the `.persistence` parameters, does it consume the default storageclass correctly? Is there a default size?

I would also break this up for readability:
```markdown
helm install mongodb \
    --set architecture=replicaset \
    --set replicaSetName=mongod \
    --set replicaCount=3 \
etc etc
```

File: docs/partners/vmware/index.md [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+    bitnami/mongodb
+```
+
+Verify that the MongoDB application has been deployed. Wait for pods to start running and PVCs to be created for each replica.
+
+```markdown
+kubectl rollout status sts/mongodb
+```
+
+Inspect the pods and PersistentVolumeClaims.
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:405]: Should we make these fixed width? `Pods` and `PersistentVolumeClaims`?

File: docs/partners/vmware/index.md [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+datadir-mongodb-1   Bound    pvc-a3755dbe-210d-4c7b-8ac1-bb0607a2c537   50Gi       RWO            primera-default-sc   13m
+datadir-mongodb-2   Bound    pvc-22bab0f4-8240-48c1-91b1-3495d038533e   50Gi       RWO            primera-default-sc   13m
+```
+
+To interact with the Mongo replica set, you can connect to the StatefulSet.
+
+```markdown
+kubectl exec -it sts/mongod bash
+
+root@mongod-0:/# df -h
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:425]: Would you mind just doing `df -h /bitnami/mongodb`?

File: docs/partners/vmware/index.md [hunk 22]
```diff
@@ -0,0 +1,451 @@
...
+
+##### Verify Cloud Native Storage in vSphere
+
+Verify that the volumes are now visible within the Cloud Native Storage interface by logging into the vSphere Web Client.
+
+Click on **Datacenter**, then the **Monitor** tab. Expand **Cloud Native Storage** and highlight **Container Volumes**. 
+
+From here, we can see the persistent volumes that were created as part of our MongoDB deployment. These should match the `kubectl get pvc` output from earlier. You can also monitor their storage policy compliance status.
+
+![Container Volumes](img/container_volumes.png)
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:449]: This is a great overview screenshot, is possible to try resizing columns and areas to show it a tad bit bigger?

**Review by datamattsson — CHANGES_REQUESTED** [review]
A few more nits and I think we should do one more pass on the install so we're not referencing deprecated bits.

File: docs/partners/vmware/index.md [hunk 23]
```diff
@@ -0,0 +1,462 @@
...
+clusterrole.rbac.authorization.k8s.io/system:cloud-controller-manager created
+
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.2.1/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml
+clusterrolebinding.rbac.authorization.k8s.io/system:cloud-controller-manager created
+
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.2.1/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml
+serviceaccount/cloud-controller-manager created
+daemonset.extensions/vsphere-cloud-controller-manager created
+service/vsphere-cloud-controller-manager created
+```
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:218]: I didn't catch this before, but could you just enumerate the `kubectl apply` and skip the output?

File: docs/partners/vmware/index.md [hunk 24]
```diff
@@ -0,0 +1,462 @@
...
+
+Verify `vsphere-cloud-controller-manager` is running.
+
+```markdown
+kubectl rollout status ds/vsphere-cloud-controller-manager -n kube-system
+```
+
+The output is similar to this:
+```markdown
+kubectl rollout status ds/vsphere-cloud-controller-manager -n kube-system
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:230]: Duplicate of what is being conveyed above.

File: docs/partners/vmware/index.md (hunk 17 above)
Comment (datamattsson) [reply docs/partners/vmware/index.md:266]: Still two backticks.

File: docs/partners/vmware/index.md [hunk 25]
```diff
@@ -0,0 +1,462 @@
...
+
+vSphere CSI node `Daemonset`:
+
+```markdown fct_label="vSphere 6.7 U3"
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.1/vsphere-67u3/vanilla/deploy/vsphere-csi-node-ds.yaml
+```
+
+```markdown fct_label="vSphere 7.0"
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.1/vsphere-7.0/vanilla/deploy/vsphere-csi-node-ds.yaml
+```
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:314]: This looks really messy. Could we summarize, like:

Deploy the necessary RBAC, vSphere CSI Controller `Deployment` and vSphere CSI node `DaemonSet`:

```
One fenced code tab chunk
```

File: docs/partners/vmware/index.md [hunk 26]
```diff
@@ -0,0 +1,462 @@
...
+
+To interact with the Mongo replica set, you can connect to the StatefulSet.
+
+```markdown
+kubectl exec -it sts/mongod bash
+
+root@mongod-0:/# df -h /bitnami/mongodb
+Filesystem      Size  Used Avail Use% Mounted on
+/dev/sdb         49G  374M   47G   1% /bitnami/mongodb
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:447]: This change has me puzzled, what shows up in the guest OS, a multipath device or a disk?

**Review by datamattsson — CHANGES_REQUESTED** [review]
Moooar nits!

File: docs/partners/vmware/index.md (hunk 26 above)
Comment (c-snell) [reply docs/partners/vmware/index.md:447]: It shows up as a disk rather than a multipath device

Comment (datamattsson) [reply docs/partners/vmware/index.md:421]: but why was it a /dev/mapper path previously?

File: docs/partners/vmware/index.md [hunk 27]
```diff
@@ -279,37 +269,21 @@ vsphere-config-secret   Opaque   1      43s
...
 kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.1/vsphere-67u3/vanilla/rbac/vsphere-csi-controller-rbac.yaml
-```
-
-```markdown fct_label="vSphere 7.0"
-kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/master/manifests/v2.0.1/vsphere-7.0/vanilla/rbac/vsphere-csi-controller-rbac.yaml
-```
-
-##### Deploy the vSphere CSI Driver
-
-vSphere CSI Controller `Deployment`:
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:276]: The extra spaces hurts a bit, could we just have the lines one after another?

File: docs/partners/vmware/index.md [hunk 28]
```diff
@@ -0,0 +1,436 @@
...
+  name: primera-default-sc
+  annotations:
+    storageclass.kubernetes.io/is-default-class: "true"
+provisioner: csi.vsphere.vmware.com
+parameters:
+  storagepolicyname: "primera-default-profile"
+  fstype: xfs
+```
+
+### Validate the vSphere CSI Driver
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:371]: If you look at the left menu how this renders, it looks a bit off. Could we just name this section "Validation""?

**Review by datamattsson — CHANGES_REQUESTED** [review]
Squash one and two more pops up.

File: docs/partners/index.md [hunk 29]
```diff
@@ -1,4 +1,6 @@
 # Partner Ecosystems
 
+* [Kasten](kasten/index.md)
 * [Rancher Labs](rancher_labs/index.md)
 * [Red Hat OpenShift](redhat_openshift/index.md)
+* [VMware](vmware/index.html)
```
Comment (datamattsson) [inline docs/partners/index.md:6]: index.md

File: docs/partners/vmware/index.md [hunk 30]
```diff
@@ -0,0 +1,430 @@
+Cloud Native Storage for vSphere
+
```
Comment (datamattsson) [inline docs/partners/vmware/index.md:2]: These lines looks a bit out of place.

**Review by datamattsson — DISMISSED** [review]
Two more nits then LGTM.

**Pull request merged**